### PR TITLE
More descriptive error message when workflow is terminated due to size limits

### PR DESCRIPTION
**Pull request by prathyushpv**
**What changed?**
Adding separate termination reasons when a workflow is terminated due to history size or history events count limit exceeded.

**Why?**
When a workflow is terminated because of history size limit or history count limit exceeded, server is not specifying which limit has exceeded and what was the actual usage in the response. This detail is just printed to the log. If the client does not have access to server logs, they will not be able to find which limit is exceeded.

**How did you test it?**
Existing test in tests/sizelimit_test.go for history count limit. Added a new test for history size limit.

**Potential risks**
None

**Is hotfix candidate?**
No

### CODE REVIEW

**Review by yiminc — COMMENTED** [review]
Please add a test for this.

**Comment by prathyushpv** [comment]
Added verification in test to check that the termination reason has history size info.

File: service/history/workflow/context.go [hunk 1]
```diff
@@ -894,10 +880,14 @@ func (c *ContextImpl) maxHistorySizeExceeded(shardContext shard.Context) bool {
 			tag.WorkflowRunID(c.workflowKey.RunID),
 			tag.WorkflowHistorySize(historySize),
 			tag.WorkflowEventCount(historyCount))
-
-		return true
+		if err := c.forceTerminateWorkflow(ctx, shardContext, fmt.Sprintf(
```
Comment (pdoerner) [inline service/history/workflow/context.go:883]: I have a minor performance concern about using `fmt.Sprintf`. Previously I have been advised against using it on hot code paths, but maybe it is okay here.

File: service/history/workflow/context.go [hunk 2]
```diff
@@ -894,10 +880,14 @@ func (c *ContextImpl) maxHistorySizeExceeded(shardContext shard.Context) bool {
...
+			common.FailureReasonHistorySizeExceedsLimit+" historySize: %v, historySizeLimit: %v, historyCount: %v, historyCountLimit: %v",
```
[the remark this answers is not in the document]
Comment (prathyushpv) [reply service/history/workflow/context.go]: Yeah that makes sense. I have updated the PR with separate failure reasons.

File: service/history/workflow/context.go [hunk 3]
```diff
@@ -938,18 +918,21 @@ func (c *ContextImpl) maxMutableStateSizeExceeded() bool {
 			tag.WorkflowRunID(c.workflowKey.RunID),
 			tag.WorkflowMutableStateSize(mutableStateSize))
 
-		return true
+		if err := c.forceTerminateWorkflow(ctx, shardContext, fmt.Sprintf(
+			common.FailureReasonMutableStateSizeExceedsLimit+" mutableStateSize: %v, mutableStateSizeLimit: %v",
```
[the remark this answers is not in the document]
Comment (prathyushpv) [reply service/history/workflow/context.go]: Got it. Changed the error message.

File: service/history/workflow/context.go (hunk 1 above)
Comment (prathyushpv) [reply service/history/workflow/context.go]: fmt.Sprintf is no longer needed.

**Review by pdoerner — APPROVED** [review]

**Pull request merged**